### PR TITLE
refactor ident

### PR DESCRIPTION
**Pull request by siddontang**
origin identifier checking and getting is complicated, slow and hard to know.
MySQL has many rules to check identifier validation like whether is ambiguous or unknown in different clauses like from, where, group by, having, select field, and order by, and we don't fully know this before and have no enough comments/docs to say why we should do this or do that. 
The origin GetIdent* functions will travel result fields every time for same identifier name, which deducts performance too.
We have use many flags to check where an identifier is, which is very complicated and hard to maintain. 

How to improve this?

+ Add refer scope and refer index field for identifier, every identifier, can only reference in select field, in from table or outer query, we can know an identifier's scope and index in Plan. so later, we can use index and scope to retrieve the value for this identifier directly. 
+ Use visitor to fix identifier scope and index in rsets group by, order by, having, etc... so we don't need any other extra checking and getting when plan executing. 
+ Learn MySQL rules again, and write detailed comments to let other people know what we don't to support. 

TODO later:
+ Check refer index validation when evaluation to prevent panic. 
+ Support referencing outer query after we can support visiting statement, and expression together. 
+ Remove unnecessary flag and function in field package for result field. 
+ Add more tests for group by, order by, having, etc...

### CODE REVIEW

File: expression/ident_test.go [hunk 1]
```diff
@@ -55,4 +55,14 @@ func (s *testIdentSuite) TestIdent(c *C) {
 	v, err = e.Eval(nil, m)
 	c.Assert(err, IsNil)
 	c.Assert(v, Equals, 1)
+
+	delete(m, ExprEvalIdentFunc)
+	e.ReferScope = 1
```
Comment (qiuyesuifeng) [inline expression/ident_test.go]: Use const variable.

File: plan/plans/groupby.go [hunk 2]
```diff
@@ -132,12 +128,19 @@ func (r *GroupByDefaultPlan) getFieldValueByName(name string, out []interface{})
...
-		if err == nil {
-			return v, nil
+	m[expression.ExprEvalIdentReferFunc] = func(name string, scope int, index int) (interface{}, error) {
+		if scope == expression.IdentReferFromTable {
+			return in[index], nil
+		} else if scope == expression.IdentReferSelectList {
+			return out[index], nil
 		}
 
+		// // TODO: remove following getting later.
```
Comment (qiuyesuifeng) [inline plan/plans/groupby.go]: Directly remove?

File: plan/plans/join.go [hunk 3]
```diff
@@ -356,32 +359,35 @@ func (r *JoinPlan) Close() error {
...
 	fields []*field.ResultField
 	row    []interface{}
+	offset int
 }
 
-// NewIdentEvalVisitor creates a new IdentEvalVisitor.
-func NewIdentEvalVisitor(fields []*field.ResultField, row []interface{}) *IdentEvalVisitor {
-	iev := &IdentEvalVisitor{fields: fields, row: row}
+// newJoinIdentVisitor creates a new joinIdentVisitor.
+func newJoinIdentVisitor(fields []*field.ResultField, offset int) *joinIdentVisitor {
```
Comment (qiuyesuifeng) [inline plan/plans/join.go]: Seems fields is useless in joinIdentVisitor?

File: plan/plans/select_list.go [hunk 4]
```diff
@@ -95,54 +95,33 @@ func (s *SelectList) GetFields() []*field.ResultField {
...
-	// We must add aggregate function to hidden select list
-	// and use a position expression to fetch its value later.
-	exprName := expr.String()
-	idx := field.GetResultFieldIndex(exprName, s.ResultFields, field.CheckFieldFlag)
-	if len(idx) == 0 {
+	if index == -1 {
 		f := &field.Field{Expr: expr}
-		resultField := &field.ResultField{Name: exprName}
-		s.AddField(f, resultField)
+		s.AddField(f, nil)
```
Comment (qiuyesuifeng) [inline plan/plans/select_list.go]: Any need to check hidden fields, like `select * from t order by count(c1) + count(c1);` seems be set in hidden fields twice?

Comment (siddontang) [inline plan/plans/select_list.go]: only one, we will only add whole order by item to hidden field, so here it is `count(c1) + count(c1)`

File: rset/rsets/helper.go [hunk 5]
```diff
@@ -34,3 +36,145 @@ func HasAggFields(fields []*field.Field) bool {
...
+		return "having clause"
+	}
+	return "none"
+}
+
+// castPosition returns an group/order by Position expression if e is a number.
+func castPosition(e expression.Expression, selectList *plans.SelectList, clause clauseType) (*expression.Position, error) {
+	v, ok := e.(expression.Value)
+	if !ok {
+		return nil, nil
```
Comment (qiuyesuifeng) [inline rset/rsets/helper.go]: Any need to return an error if expression is not expression.Value？

File: rset/rsets/having.go [hunk 6]
```diff
@@ -31,66 +31,266 @@ type HavingRset struct {
...
+func (v *havingVisitor) VisitIdent(i *expression.Ident) (expression.Expression, error) {
+	// Having has the most complex rule for ambiguous and identifer reference check.
+
+	// Having ambiguous rule:
+	//	select c1 as a, c2 as a from t having a is ambiguous
+	//	select c1 as a, c2 as a from t having a + 1 is ambiguous
+	//	select c1 as c2, c2 from t having c2 is ambiguous
+	//	select c1 as c2, c2 from t having c2 + 1 is ambiguous
+
+	// The identifier in having must exist in group by, select list and outer query, so select c1 from t having c2 is wrong,
```
Comment (coocood) [inline rset/rsets/having.go]: group by, select list ***or*** outer query

File: rset/rsets/having.go [hunk 7]
```diff
@@ -31,66 +31,266 @@ type HavingRset struct {
...
+	// the idenfitier will first try to find the reference in group by, then in select list and finally in outer query.
```
Comment (coocood) [inline rset/rsets/having.go]: idenfitier -> identifier

File: rset/rsets/helper.go (hunk 5 above)
Comment (siddontang) [inline rset/rsets/helper.go]: no need, you can see the value that we will use another check if return nil.

File: expression/helper.go [hunk 8]
```diff
@@ -142,22 +144,40 @@ func newMentionedAggregateFuncsVisitor() *mentionedAggregateFuncsVisitor {
...
+		// here means we have aggregate function in arg.
+		return nil, errors.Errorf("Invalid use of group function")
 	}
+
 	return c, nil
 }
 
+// IsAggregateFunc checks whether name is an aggregate function or not.
+func IsAggregateFunc(name string) (bool, error) {
+	f, ok := builtin.Funcs[strings.ToLower(name)]
```
Comment (shenli) [inline expression/helper.go]: Some of the builtin functions are not in the builtin.Funcs table.

**Comment by shenli** [comment]
LGTM

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**